File: app/schemas/highlight.py
import uuid
from datetime import datetime
from pydantic import BaseModel


class AncestorNode(BaseModel):
    """Heading ancestor entry stored with each highlight for summary reconstruction."""
    node_id: str
    level: int
    text: str


class HighlightCreate(BaseModel):
    node_ids: list[str]                  # IDs of selected StandardFormat nodes
    color: str = "yellow"
    note: str | None = None


class HighlightRead(BaseModel):
    id: uuid.UUID
    document_id: uuid.UUID
    node_ids: list[str]
    ancestor_node_ids: list[str]
    note: str | None = None
    color: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class HighlightUpdate(BaseModel):
    note: str | None = None
    color: str | None = None


class SummaryView(BaseModel):
    """
    Derived view: the subset of a document's Standard Format that falls
    under highlighted content, including ancestor heading chains.
    """
    document_id: uuid.UUID
    document_title: str
    sections: list[dict]   # Reconstructed tree segments with heading context

File: packages/shared/src/types/highlight.ts
export type HighlightColor = "yellow" | "green" | "blue" | "pink" | "purple";

export interface Highlight {
  id: string;
  document_id: string;
  node_id: string;
  start_offset?: number;
  end_offset?: number;
  ancestor_headings: string[];
  color: HighlightColor;
  note?: string;
  created_at: string;
}

export interface HighlightCreate {
  node_id: string;
  start_offset?: number;
  end_offset?: number;
  color: HighlightColor;
  note?: string;
}

export interface HighlightUpdate {
  color?: HighlightColor;
  note?: string;
}

export interface SummarySection {
  ancestor_headings: string[];
  text: string;
  highlight_id: string;
  color: HighlightColor;
  note?: string;
}
